lib/semantics/check-allocate.cc

```diff
@@ -0,0 +1,411 @@
+  bool gotStat{false};
+  bool gotMsg{false};
+  bool gotTypeSpec{false};
+  bool gotSrc{false};
+  bool gotMold{false};
+};
+
+static const parser::Name &GetName(
+    const parser::AllocateObject &allocateObject) {
+  return std::visit(
+      common::visitors{
+          [&](const parser::Name &name) -> const parser::Name & {
+            return name;
+          },
+          [&](const parser::StructureComponent &structureComponent)
+              -> const parser::Name & { return structureComponent.component; },
+      },
+      allocateObject.u);
+}
+
+class AllocationChecker {
+public:
+  AllocationChecker(
+      const parser::AllocateObject &obj, AllocateCheckerInfo &info)
+    : allocateInfo_{info}, name_{GetName(obj)}, type_{name_.symbol->GetType()},
+      isSubobject_{std::holds_alternative<parser::StructureComponent>(obj.u)} {
+    CHECK(type_ != nullptr);
+    if (type_->category() == DeclTypeSpec::Category::Character) {
+      hasDefferedTypeParameter_ =
+          type_->characterTypeSpec().length().isDeferred();
+    } else if (const DerivedTypeSpec * derivedTypeSpec{type_->AsDerived()}) {
+      for (auto it{derivedTypeSpec->parameters().begin()};
+           it != derivedTypeSpec->parameters().end(); ++it) {
+        hasDefferedTypeParameter_ |= it->second.isDeferred();
+      }
+      isAbstract_ = derivedTypeSpec->typeSymbol().attrs().test(Attr::ABSTRACT);
+    }
+    isUnlimitedPolymorphic_ =
+        type_->category() == DeclTypeSpec::Category::ClassStar;
+  }
+
+  bool RunChecks(SemanticsContext &context);
+
+private:
+  AllocateCheckerInfo &allocateInfo_;
+  const parser::Name &name_;
+  const DeclTypeSpec *type_;
+  bool isSubobject_;
+  bool hasDefferedTypeParameter_{false};
+  bool isUnlimitedPolymorphic_{false};
+  bool isAbstract_{false};
+};
+
+static std::optional<AllocateCheckerInfo> CheckAllocateOptions(
+    const parser::AllocateStmt &, SemanticsContext &);
+static bool IsTypeCompatible(const DeclTypeSpec &, const DeclTypeSpec &);
+static bool IsTypeCompatible(
+    const DeclTypeSpec &type1, const evaluate::DynamicType &type2);
+static bool HaveSameAssumedTypeParameters(
+    const DeclTypeSpec &, const DeclTypeSpec &);
+static bool HaveCompatibleKindParameters(
+    const DeclTypeSpec &, const DeclTypeSpec &);
+
+void AllocateChecker::Leave(const parser::AllocateStmt &allocateStmt) {
+  if (auto info{CheckAllocateOptions(allocateStmt, context_)}) {
+    for (const parser::Allocation &allocation :
+        std::get<std::list<parser::Allocation>>(allocateStmt.t)) {
+      AllocationChecker allocationChecker{
+          std::get<parser::AllocateObject>(allocation.t), *info};
+      allocationChecker.RunChecks(context_);
+    }
+  }
+}
+
+static std::optional<AllocateCheckerInfo> CheckAllocateOptions(
+    const parser::AllocateStmt &allocateStmt, SemanticsContext &context) {
+  AllocateCheckerInfo info;
+  evaluate::ExpressionAnalyzer analyzer{context};  // can emit error messages
```

Comment:
I don't think it's necessary to analyze the Exprs and Variables -- ExprChecker should have done it already.

Reply:
OK, I took this from other check-xx.cc. I saw that you removed these since and I could not find tests to "use" this anyway. Removed as suggested.